community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/AbstractKeyValueStore.java

```diff
@@ -300,14 +300,27 @@ private long rotate( boolean force ) throws IOException
         {
             try( RotationState<Key> rotation = this.rotation )
             {
-                final long version = rotation.rotationVersion();
-                ProgressiveState<Key> next = rotation.rotate( force, rotationStrategy, rotationTimerFactory,
-                        value -> updateHeaders( value, version ) );
-                try ( LockWrapper ignored = writeLock( updateLock ) )
+                try
                 {
-                    state = next;
+                    final long version = rotation.rotationVersion();
+                    ProgressiveState<Key> next = rotation.rotate( force, rotationStrategy, rotationTimerFactory,
+                            value -> updateHeaders( value, version ) );
+                    try ( LockWrapper ignored = writeLock( updateLock ) )
+                    {
+                        state = next;
+                    }
+                    return version;
+                }
+                catch ( Throwable t )
+                {
+                    // Rotation failed. Here we assume that rotation state remembers this so that closing it
+                    // won't close the state as it was before rotation began, which we're reverting to right here.
```

Comment:
The test failure was due to shutdown not being able to close the state (since it was in rotating state when shutdown() was called. This leaves the rotation task the only guy that can close it. Made a change there so that it knows when it's left with this responsibility.